Fix ILLink warnings for removed attributes when Debugger.IsSupported is false

eerhardt opened the pull request:
When `Debugger.IsSupported` feature switch is false, we trim unnecessary attributes in the code - specifically attributes like `CompilerGeneratedAttribute` and `StackTraceHiddenAttribute`. However, we still have code accessing these attributes, which causes the ILLinker to warn that it is removing the attributes, but code still uses them.

To fix this, remove these attributes from being trimmed when Debugger.IsSupported is false. Instead these attributes will be preserved always so things like "is anonymous type" checks and stack trace tostring will behave correctly.

Fix #39707

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/Statics.cs:
```diff
@@ -503,7 +503,7 @@ public static Delegate CreateDelegate(Type delegateType, MethodInfo methodInfo)
             EventDataAttribute? eventAttrib = Statics.GetCustomAttribute<EventDataAttribute>(dataType);
             if (eventAttrib != null ||
-                Statics.GetCustomAttribute<CompilerGeneratedAttribute>(dataType) != null ||
+                IsCompilerGenerated(dataType) ||
```
If I am reading this correctly, turning off Debugger support would lead to silent behavior changes in Tracing. It is unpredictable experience.

eerhardt replied:
This change isn't critical since EventSource is disabled by default in Blazor. I included it for completeness, but I can undo it. The resulting behavior would be if someone trimmed their app with `Debugger.IsSupported=false` and `EventSource.IsSupported=true`, they would still get `CompilerGeneratedAttributes` in their binaries (after https://github.com/mono/linker/issues/1378 is fixed). So the size would grow slightly, but `EventSource` would work the exact same.

As an aside, I'm a little bit confused by the existing code here. When would someone pass a compiler generated type into `EventSource.Write`? Is this how we check for anonymous types?

cc @brianrob @noahfalk

vitek-karas replied:
If we don't need I agree we should leave it as-is for now.

MichalStrehovsky replied:
> Is this how we check for anonymous types?

Yes, we document that in the EventSource.Write docs: "The event data. This type must be an anonymous type or marked with the EventDataAttribute attribute." and this check is trying to do that. There isn't a better way to check for anonymous types that I know of.

vargaz commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/StackTrace.cs:
```diff
@@ -225,12 +225,19 @@ internal void ToString(TraceFormat traceFormat, StringBuilder sb)
+                    // So only access CompilerGeneratedAttribute when Debugger.IsSupported is true.
+                    if (Debugger.IsSupported)
                     {
-                        isAsync = typeof(IAsyncStateMachine).IsAssignableFrom(declaringType);
```
Won't this change the stacktrace format ?

eerhardt replied:
Correct, it does. See https://github.com/dotnet/runtime/pull/39000#discussion_r452362679 for discussion.

MichalStrehovsky commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/StackTrace.cs:
```diff
@@ -225,12 +225,19 @@ internal void ToString(TraceFormat traceFormat, StringBuilder sb)
                     bool methodChanged = false;
-                    if (declaringType != null && declaringType.IsDefined(typeof(CompilerGeneratedAttribute), inherit: false))
+
+                    // Removing Debugger support with the ILLinker will remove CompilerGeneratedAttribute.
```
I'm a little bit conflicted by this. Up until now, `System.Diagnostics.Debugger.IsSupported` feature basically meant - this is the stuff that is necessary to either attach a debugger at runtime, or to analyze crash dumps post mortem. We have scenarios where removing this support clearly makes sense (WASM, and probably also Xamarin, since (managed) dump debugging where one would be able to use debugger visualizers with Mono is non-existent as far as I know).

But this is about decoding async stack traces that are observable at runtime and are unrelated with the debugging scenarios above. This is the stack traces that people use to implement things like crash telemetry.

Do we really want to lump all of this together? I would also question making this the default now without enough time to get feedback.

noahfalk replied:
Unless it is very onerous to add additional checks I'd recommend not lumping this together. Agree with @MichalStrehovsky that stack traces show up in logs, telemetry, Exceptions, and crash reporting, none of which a customer is likely to think about if we say "this scenario doesn't support a debugger."

eerhardt replied:
cc @marek-safar who originally proposed removing `CompilerGeneratedAttribute` with https://github.com/dotnet/runtime/pull/39000#discussion_r452362679.

I would be fine with never trimming `CompilerGeneratedAttribute` and `StackTraceHiddenAttribute`. `CompilerGeneratedAttribute` provides a modest size savings in my investigation: roughly `11 KB` uncompressed IL, when compressed the savings are even smaller.

eerhardt committed:
Fix ILLink warnings for removed attributes when Debugger.IsSupported is false

When Debugger.IsSupported feature switch is false, we trim unnecessary attributes in the code - specifically attributes like CompilerGeneratedAttribute and StackTraceHiddenAttribute. However, we still have code accessing these attributes, which causes the ILLinker to warn that it is removing the attributes, but code still uses them.

To fix this, remove these attributes from being trimmed when Debugger.IsSupported is false. Instead these attributes will be preserved always so things like "is anonymous type" checks and stack trace tostring will behave correctly.

Fix #39707

eerhardt commented:
Thanks everyone for the feedback. I am now removing `CompilerGeneratedAttribute` and `StackTraceHiddenAttribute` from the list of attributes being trimmed by the linker when Debugger.IsSupported is false. These attributes will always be preserved in the assemblies.

fyi - @marek-safar

marek-safar commented:
@eerhardt what is the negative size impact of this change for Blazor and how are we going to mitigate it?

eerhardt commented:
> what is the negative size impact of this change for Blazor 

|            | Uncompressed IL        | Compressed IL (.br)  |
|------------|------------------------|----------------------|
| master     | 2,412,032 bytes        | 847,333 bytes        |
| PR         | 2,423,808 bytes        | 850,780 bytes        |
| Difference | 11,776 bytes (11.5 KB) | 3,447 bytes (3.4 KB) |

> and how are we going to mitigate it?

I am not sure it is necessary to mitigate it. 3 KB compressed is a pretty minimal change when the total compressed app size is around 2.2 MB.

eerhardt commented:
Test failures are #40178. I believe this is ready to merge. Any other feedback?
